flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/PartialDispatcherServices.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.dispatcher;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.blob.BlobServer;
+import org.apache.flink.runtime.heartbeat.HeartbeatServices;
+import org.apache.flink.runtime.highavailability.HighAvailabilityServices;
+import org.apache.flink.runtime.metrics.groups.JobManagerMetricGroup;
+import org.apache.flink.runtime.resourcemanager.ResourceManagerGateway;
+import org.apache.flink.runtime.rpc.FatalErrorHandler;
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Partial {@link DispatcherServices} services container which needs to
+ * be completed before being given to the {@link Dispatcher}.
+ */
+public class PartialDispatcherServices {
```

Comment:
Actually there is a similar discuss [here](https://github.com/apache/flink/pull/9787).

Reply:
What would be your suggestion for a better name @buptljy? I agree that `DispatcherServices` does not need to strictly inherit from `PartialDispatcherServices`.

Reply:
The intention of the `PartialDispatcherServices` and `DispatcherServices` is to be a container for the dispatcher related services.

Reply:
With an offline discussion with @buptljy it happens raise that an idea that

```java
public final class DispatcherServicesConstructionHelper { ... }

public /* maybe final */class DispatcherService {
  // ...
  public static DispatcherService from(DispatcherServicesConstructionHelper, JMFactory) { ... }
}
```

for help a bit on naming issue and prevent inheritance leak.

Reply:
To clarify inherit from `DispatcherService` might be OK for testing but from `PartialDispatcherServices` looks error prone.

Reply:
Still I am thinking of `DispatcherServices` for sharing services between different Dispatcher leader epoch since we possibly separate concern of leader election and job management from current Dispatcher and let `DispatcherRunner` be the supervisor or `Dispatcher` that start a new instance per leader epoch.

Reply:
Fair enough, I'll remove the inheritance. 

Concerning the renaming, I would like to postpone it until I have merged the remaining PRs in order to fix FLINK-11843 and then do the renaming because otherwise I would end up with quite a bit of merge conflicts.

Concerning the lifecycle management of `DispatcherServices` I think this should be another issue. As I've said, the intention is not to introduce such a lifecycle management but simply to have a service container.

Reply:
Thanks for sharing your thoughts. You are the coordinator of these batch of pull requests. For concerns of lifecycle management let's move to another thread. I mention here to help express my perspective about the name and inheritance :-) It is simply a service container in this series of patches, definitely.